pkg/standalone-utils/contracts/relayer/VaultActions.sol

```diff
@@ -280,4 +330,108 @@ abstract contract VaultActions is IBaseRelayerLibrary {
             return userData;
         }
     }
+
+    function _doLegacyStableExitChainedReferenceReplacements(bytes memory userData, uint8 exitKind)
+        private
+        returns (bytes memory)
+    {
+        if (exitKind == uint8(LegacyStablePoolUserData.ExitKind.EXACT_BPT_IN_FOR_ONE_TOKEN_OUT)) {
+            return _doStableExactBptInForOneTokenOutReplacements(userData, exitKind);
+        } else if (exitKind == uint8(LegacyStablePoolUserData.ExitKind.EXACT_BPT_IN_FOR_TOKENS_OUT)) {
+            return _doStableExactBptInForTokensOutReplacements(userData, exitKind);
+        } else {
+            // All other exit kinds are 'given out' (i.e the parameter is a token amount),
+            // so we don't do replacements for those.
+            return userData;
+        }
+    }
+
+    function _doLegacyComposableStableExitChainedReferenceReplacements(bytes memory userData, uint8 exitKind)
+        private
+        returns (bytes memory)
+    {
+        if (exitKind == uint8(LegacyComposableStablePoolUserData.ExitKind.EXACT_BPT_IN_FOR_ONE_TOKEN_OUT)) {
+            return _doStableExactBptInForOneTokenOutReplacements(userData, exitKind);
+        } else {
+            // All other exit kinds are 'given out' (i.e the parameter is a token amount),
+            // so we don't do replacements for those.
+            return userData;
+        }
+    }
+
+    // For ComposableStablePool V2 and V3
+    function _doComposableStableExitChainedReferenceReplacements(bytes memory userData, uint8 exitKind)
+        private
+        returns (bytes memory)
+    {
+        if (exitKind == uint8(StablePoolUserData.ExitKind.EXACT_BPT_IN_FOR_ONE_TOKEN_OUT)) {
+            return _doStableExactBptInForOneTokenOutReplacements(userData, exitKind);
+        } else if (exitKind == uint8(StablePoolUserData.ExitKind.EXACT_BPT_IN_FOR_ALL_TOKENS_OUT)) {
+            return _doStableExactBptInForTokensOutReplacements(userData, exitKind);
+        } else {
+            // All other exit kinds are 'given out' (i.e the parameter is a token amount),
+            // so we don't do replacements for those.
+            return userData;
+        }
+    }
+
+    // Shared Stable Exit utility functions
+
+    /**
+     * @dev Can use the standard library for parsing out the arguments, as this is common to all types, but we need
+     * to pass in the exitKind itself, as this could be different.
```

Comment:
I think we need to have a note somewhere (perhaps when choosing stable pool kinds) that explains that they don't all support the same exit kinds, and that in some cases the encoding is the same but the exit kind id has different values, so we handle them this way.

It's otherwise a bit confusing to follow, and close to impossible if you don't already know that.